Forte.EpiServer.AzureSearch/Events/SearchEventHandler.cs

```diff
@@ -1,95 +1,158 @@
 namespace Forte.EpiServer.AzureSearch.Events
 {
     public class SearchEventHandler<T> where T : ContentDocument
     {
         private readonly IAzureSearchService _azureSearchService;
-        private readonly IContentDocumentBuilder<T> _contentDocumentBuilder;
         private readonly IContentLoader _contentLoader;
+        private readonly PageSearchEventHandler<T> _pageSearchEventHandler;
+        private readonly BlockSearchEventHandler<T> _blockSearchEventHandler;
+        private readonly IUrlResolver _urlResolver;
 
-        public SearchEventHandler(IAzureSearchService azureSearchService, IContentDocumentBuilder<T> contentDocumentBuilder, IContentLoader contentLoader)
+        public SearchEventHandler(IAzureSearchService azureSearchService, IContentLoader contentLoader,
+            PageSearchEventHandler<T> pageSearchEventHandler, BlockSearchEventHandler<T> blockSearchEventHandler,
+            IUrlResolver urlResolver)
         {
             _azureSearchService = azureSearchService;
-            _contentDocumentBuilder = contentDocumentBuilder;
             _contentLoader = contentLoader;
+            _pageSearchEventHandler = pageSearchEventHandler;
+            _blockSearchEventHandler = blockSearchEventHandler;
+            _urlResolver = urlResolver;
         }
 
-        public void OnPublishedContent(object sender, ContentEventArgs contentEventArgs)
+        //Event handler for keeping 'OldUrl' in contentEventArgs
+        public void OnPublishingContent(object sender, ContentEventArgs contentEventArgs)
         {
-            if (!contentEventArgs.Content.ShouldIndex())
+            if (contentEventArgs.Content is PageData == false)
             {
                 return;
             }
-            
-            var contentInAllLanguageVersions = _contentLoader.GetAllLanguageVersions(contentEventArgs.Content.ContentLink);
-            var documents = contentInAllLanguageVersions.Select(c => _contentDocumentBuilder.Build(c));
-            
-            Task.Run(() => _azureSearchService.IndexAsync(documents));
+            var oldUrl = _urlResolver.GetUrl(new ContentReference(contentEventArgs.Content.ContentLink.ID));
+
+            if (string.IsNullOrEmpty(oldUrl) == false)
+            {
+                contentEventArgs.Items.Add(PageSearchEventHandler<T>.OldUrlKey, oldUrl);
+            }
+        }
+        
+        //Event handler for publishing page and block content
+        public void OnPublishedContent(object sender, ContentEventArgs contentEventArgs)
+        {
+            var content = contentEventArgs.Content;
+
+            switch (content)
+            {
+                case PageData _:
+                    var documentsToIndex = _pageSearchEventHandler.GetDocuments(content, contentEventArgs);
+                    Task.Run(() => _azureSearchService.IndexAsync(documentsToIndex));
+                    break;
+
+                case BlockData _:
+                    UpdateBlockParentPagesInIndex(content);
```

Comment:
It's tricky that you sometimes run task in this method "OnPublishedContent" and sometimes inside a method that you call (i.e. UpdateBlockParentPagesInIndex). 

Reading this code I would assume that UpdateBlockParentPagesInIndex() is always synchronous. Maybe you should move Task.Run() to this method as well?

Another option is to be more explicit in method names (naming is a solution for everything).

E.g. ScheduleUpdatingBlockParentPagesInIndex() or something similar